### PR TITLE
feat: add search encoder backend

### PR DESCRIPTION
**Pull request by Samoed**
Close https://github.com/embeddings-benchmark/mteb/issues/3406 

I've implemented SearchEncoder protocol and now can be selected faiss or direct search (as previous). 

https://github.com/embeddings-benchmark/mteb/blob/c8b2bd3a5d1af6a8639da567d6dfd30cbbd25bdb/mteb/models/search_encoder_index/search_backend_protocol.py#L7-L29

I've saved "batched" approach for retrieval to store less memory during evaluation. Backend can be changed by 

```python
import mteb
from mteb.models.search_encoder_index import (
    DefaultEncoderSearchBackend,
    FaissEncoderSearchBackend,
)
from mteb.models import SearchEncoderWrapper


model = mteb.get_model("baseline/random-encoder-baseline")

python_backend = SearchEncoderWrapper(
    model, index_backend=DefaultEncoderSearchBackend()
)
faiss_backend = SearchEncoderWrapper(
    model, index_backend=FaissEncoderSearchBackend(model)
)
```

I've tested on Scifact using `potion-2M` and got 2s evaluation for default search and 3s for `FAISS`.

<details>
<summary>Script to test</summary>

```python
import mteb
from mteb.cache import ResultCache
from mteb.models import SearchEncoderWrapper
from mteb.models.search_encoder_index import (
    DefaultEncoderSearchBackend,
    FaissEncoderSearchBackend,
)

model = mteb.get_model("minishlab/potion-base-2M")

python_backend = SearchEncoderWrapper(
    model, index_backend=DefaultEncoderSearchBackend()
)
faiss_backend = SearchEncoderWrapper(
    model, index_backend=FaissEncoderSearchBackend(model)
)

task = mteb.get_task("SciFact")

python_cache = ResultCache("python_backend_cache")
faiss_cache = ResultCache("faiss_backend_cache")

# warmup
mteb.evaluate(
    model,
    task,
    cache=None,
)

mteb.evaluate(
    python_backend,
    task,
    cache=python_cache,
)

mteb.evaluate(
    faiss_backend,
    task,
    cache=faiss_cache,
)
```

</details>

### CODE REVIEW

**Review by orionw — APPROVED** [review]
Looks great overall!

I think it doesn't take advantage of `faiss`'s built in scoring functionality but by not doing that we can have more control (so that can be ignored). If we wanted to use faiss we could do something like

```
# Batch reconstruct candidate embeddings
candidate_embs = np.vstack([
    self.index.reconstruct(idx) for idx in candidate_indices
])

# Create temporary index to let FAISS handle scoring
temp_index = self.index_type(d)
temp_index.add(candidate_embs)

# Search returns scores and indices in one call
scores, local_indices = temp_index.search(
    query_emb.reshape(1, -1).astype(np.float32),
    min(top_k, len(candidate_indices))
)
```

But I think it just does dot product. So it looks great as is, but just mentioning this in case that's helpful.

**Comment by Samoed** [comment]
Yes, I think that's better. I've added support of cosine and dot product similarity support and scores are nearly the same (same for `1e-6`).

File: mteb/models/search_encoder_index/search_backend_protocol.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
+from collections.abc import Callable
+from typing import Protocol
+
+from mteb.types import Array, TopRankedDocumentsType
+
+
+class IndexEncoderSearchProtocol(Protocol):
```
Comment (KennethEnevoldsen) [inline mteb/models/search_encoder_index/search_backend_protocol.py:7]:
```suggestion
class EncoderSearchProtocol(Protocol):
```

File: mteb/models/search_encoder_index/default_backend_search.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+from collections.abc import Callable
+
+import torch
+
+from mteb.types import Array, TopRankedDocumentsType
+
+logger = logging.getLogger(__name__)
+
+
+class DefaultEncoderSearchBackend:
```
Comment (KennethEnevoldsen) [inline mteb/models/search_encoder_index/default_backend_search.py]:
```suggestion
class StreamingSearchBackend:
```
Should we call it backend or "index"

File: mteb/models/search_encoder_index/faiss_search_backend.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+from mteb._requires_package import requires_package
+from mteb.models.model_meta import ScoringFunction
+from mteb.models.models_protocols import EncoderProtocol
+from mteb.types import Array, TopRankedDocumentsType
+
+logger = logging.getLogger(__name__)
+
+
+class FaissEncoderSearchBackend:
```
Comment (KennethEnevoldsen) [inline mteb/models/search_encoder_index/faiss_search_backend.py]:
```suggestion
class FaissSearchBackend:
```

File: tests/test_abstasks/test_predictions.py [hunk 4]
```diff
@@ -125,3 +134,68 @@ def test_predictions(tmp_path: Path, task, expected):
 
     predictions = full_predictions["default"]["test"]
     assert predictions == expected
+
+
+@pytest.mark.parametrize(
```
Comment (KennethEnevoldsen) [inline tests/test_abstasks/test_predictions.py]: should this be placed here? I would say that it neither under abstasks nor test_predictions

**Comment by Samoed** [comment]
> We also need to add documentation

Yes, wanted to add after your check on pr

File: mteb/models/search_encoder_index/search_backend_protocol.py (hunk 1 above)
Comment (Samoed) [reply mteb/models/search_encoder_index/search_backend_protocol.py:7]: I think we can specify that this is only for index and only for encoder, because this can be confused that `SentenceTransformerEncoderWrapper` will implement it (probably)

File: mteb/models/search_encoder_index/default_backend_search.py (hunk 2 above)
Comment (Samoed) [reply mteb/models/search_encoder_index/default_backend_search.py]: I think index would be better to use

File: mteb/models/search_encoder_index/faiss_search_backend.py (hunk 3 above)
Comment (Samoed) [reply mteb/models/search_encoder_index/faiss_search_backend.py]: Renamed to `FaissSearchIndex`

File: mteb/models/search_encoder_index/default_backend_search.py (hunk 2 above)
Comment (Samoed) [reply mteb/models/search_encoder_index/default_backend_search.py]: Renamed to `StreamingSearchIndex`

File: tests/test_abstasks/test_predictions.py (hunk 4 above)
Comment (Samoed) [reply tests/test_abstasks/test_predictions.py]: Created new folder test_search_index

Comment (KennethEnevoldsen) [reply tests/test_abstasks/test_predictions.py]: wouldn't you put it test_models (to match the structure of the repo?)

Comment (Samoed) [reply tests/test_abstasks/test_predictions.py]: Yes, will move it here then

**Comment by Samoed** [comment]
I've run this script and both evaluation method took same time, so I'm unsure a bit what to add in advantages of FAISS, except of dumping index, but we're clearing it after evaluation.

| task | Stream | FAISS |
|-|-|-|
| SWEbenchVerifiedRR | 536 | 541 |
| ClimateFEVERHardNegatives | 9 | 12 | 

```python
import logging

import mteb
from mteb.cache import ResultCache
from mteb.models import SearchEncoderWrapper
from mteb.models.search_encoder_index import StreamingSearchIndex, FaissSearchIndex

logging.basicConfig(level=logging.INFO)

model = SearchEncoderWrapper(mteb.get_model("minishlab/potion-base-2M"))
tasks = mteb.get_tasks(
    tasks=[
        "ClimateFEVERHardNegatives",
        "SWEbenchVerifiedRR",
    ],
)

cache = ResultCache("stream")

mteb.evaluate(
    model,
    tasks,
    cache=cache,
)

### FAISS
index_backend = FaissSearchIndex(model)
model = SearchEncoderWrapper(
    mteb.get_model("minishlab/potion-base-2M"),
    index_backend=index_backend
)
cache = ResultCache("FAISS")

mteb.evaluate(
    model,
    tasks,
    cache=cache,
)
```

**Comment by orionw** [comment]
I think faiss is not ideal for smaller reranking cases (~100-1000 docs to search for). We should see dramatic gains for retrieval though, with a large enough corpus. For `ClimateFEVERHardNegatives` it could just be initialization differences. Maybe if you try MS MARCO for retrieval?

I asked Claude what it thinks we should do for reranking and it suggested we retrieve the vectors from `faiss` for reranking but just use standard numpy afterwards. We could do this, but if it's roughly the same to use faiss then we might as well keep what we have for that.

If large scale retrieval is much faster I think that's the main benefit

**Comment by Samoed** [comment]
I tried running it on `MSMARCO`, and both backends showed similar times on sub-batches. If we remove the search over each sub-corpus batch, FAISS would probably show a speedup, but I’m not sure how to do that while still supporting the "streaming" backend.

File: mteb/models/search_encoder_index/search_backend_protocol.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    """Protocol for search backends used in encoder-based retrieval."""
+
+    def add_document(
```
Comment (KennethEnevoldsen) [inline mteb/models/search_encoder_index/search_backend_protocol.py]:
```suggestion
    def add_documents(
```

File: mteb/models/search_wrappers.py [hunk 6]
```diff
@@ -173,23 +181,13 @@ def _full_corpus_search(
                 prompt_type=PromptType.document,
                 **encode_kwargs,
             )
-
+            self.index_backend.add_document(sub_corpus_embeddings, sub_corpus_ids)
```
Comment (KennethEnevoldsen) [inline mteb/models/search_wrappers.py]: shouldn't things we added in the index step? (I suspect it it here because of the streaming backend)

Comment (Samoed) [reply mteb/models/search_wrappers.py]: Yes, but I don't know how to make it compatible with streaming without creating new functions. If we want to use the index step, we need duplicate functions with similar logic. Maybe we can create different wrapper to use without hacks to use index directly

Comment (KennethEnevoldsen) [reply mteb/models/search_wrappers.py]: Hmm can you outline the current approach in the streaming just to clarify why it is the way it is now (Just to make sure that we agree on the problem).

Comment (Samoed) [reply mteb/models/search_wrappers.py]: Current approach with "streaming" don't store corpus embeddings fully for retrieval. It processes corpus in "chunks" of 50_000 documents and then finds similarities between documents and queries. Because of that we're adding `sub_corpus_embeddings` here (chunk embeddings). If we move encoding of documents in index, then we need to store all embeddings of documents, but this will cause problems with large datasets like miracl or msmarco

Comment (KennethEnevoldsen) [reply mteb/models/search_wrappers.py]: Hmm, so I see two possible issues. 

1) Either the abstraction is faulty, and we should try to come up with a different protocol. E.g., one that takes in the encoder.
2) or the streaming backend doesn't fit into the current abstraction, and we should remove the streaming from it. E.g. by making `SearchEncoderWrapper` (takes a search engine) and `StreamingSearchEncoderWrapper` (we probably have to word with the naming to avoid breaking backward compat).

Comment (Samoed) [reply mteb/models/search_wrappers.py]: > Either the abstraction is faulty, and we should try to come up with a different protocol. E.g., one that takes in the encoder.

I think if we do that then it will be almost like the search protocol 

> the streaming backend doesn't fit into the current abstraction, and we should remove the streaming from it

I think yes, but I don't know how to do this without code duplication. We can make if no index backend is passed, then a search approach will be used, otherwise index

Comment (KennethEnevoldsen) [reply mteb/models/search_wrappers.py]: > I think if we do that then it will be almost like the search protocol

Yep, that was also my worry - good to see that we agree

> I think yes, but I don't know how to do this without code duplication. We can make if no index backend is passed, then a search approach will be used, otherwise index

Yeah, that might actually be a pretty decent approach

Comment (Samoed) [reply mteb/models/search_wrappers.py]: Made backend optional

File: mteb/models/search_encoder_index/search_indexes/streaming_search_index.py
Comment (KennethEnevoldsen) [inline mteb/models/search_encoder_index/search_indexes/streaming_search_index.py:1]: doesn't that mean that we should delete this?  (since it is the default)

Comment (Samoed) [reply mteb/models/search_encoder_index/search_indexes/streaming_search_index.py:1]: Removed

File: docs/advanced_usage/retrieval_backend.md [hunk 7]
```diff
@@ -0,0 +1,23 @@
+# Retrieval Search backend
+
+!!! note "Available since 2.3.0"
+    This feature was introduced in version **2.3.0**.
+
+For some large dataset search can take a lot of time. To reduce this you can use `FaissSearchIndex`. To work with it install `pip install mteb[faiss]`.
```
Comment (KennethEnevoldsen) [inline docs/advanced_usage/retrieval_backend.md]: Wasn't the main concern memory?

File: docs/advanced_usage/retrieval_backend.md
Comment (KennethEnevoldsen) [inline docs/advanced_usage/retrieval_backend.md:1]: shouldn't we also add documentation on the search backend?

Maybe we should also add something here so people can discover what has been added:

Screenshot 2025-11-25 at 16 23 28

A kind of user-friendly changelog

Comment (Samoed) [reply docs/advanced_usage/retrieval_backend.md:1]: It will be shown in `advanced usage`

Comment (KennethEnevoldsen) [reply docs/advanced_usage/retrieval_backend.md:1]: Yeah, but people will not know what has happened since 2.0.0

I would probably change `New in v2.0` to 

```
- What is new
  - v2.3
  - v2.2
  - v2.1
  - v2.0
```

Comment (Samoed) [reply docs/advanced_usage/retrieval_backend.md:1]: I think this is more about changelog https://github.com/embeddings-benchmark/mteb/issues/3401

File: docs/advanced_usage/retrieval_backend.md (hunk 7 above)
Comment (Samoed) [reply docs/advanced_usage/retrieval_backend.md]: Also added memory

File: docs/advanced_usage/retrieval_backend.md
Comment (KennethEnevoldsen) [reply docs/advanced_usage/retrieval_backend.md:1]: Fair we still need the API docs though

File: Makefile [hunk 8]
```diff
@@ -35,9 +35,8 @@ pr:
 	make lint
 	make test
 
-build-docs:
+build-docs: build-docs-overview
```
Comment (KennethEnevoldsen) [inline Makefile:38]: oO does this work?

**Comment by KennethEnevoldsen** [comment]
I think this is good to merge

File: Makefile (hunk 8 above)
Comment (Samoed) [reply Makefile:38]: Yes, everything after `:` will be triggered before running a function
```makefile
targets: prerequisites
	command
```
